src/twisted/internet/defer.py

```diff
@@ -515,54 +542,151 @@ def addCallbacks(
+        self,
+        callback: Any,
+        *args: Any,
+        **kwargs: Any,
+    ) -> "Deferred[Any]":
         """
         Convenience method for adding just a callback.
 
         See L{addCallbacks}.
         """
         return self.addCallbacks(callback, callbackArgs=args, callbackKeywords=kwargs)
 
+    @overload
     def addErrback(
         self,
-        errback: Callable[
-            ...,
-            "Union[Failure, _NextDeferredResultT, Deferred[_NextDeferredResultT]]",
-        ],
-        *args: object,
-        **kwargs: object,
+        errback: Callable[Concatenate[Failure, _P], Deferred[_NextDeferredResultT]],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
+        ...
+
+    @overload
+    def addErrback(
+        self,
+        errback: Callable[Concatenate[Failure, _P], Failure],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT]]":
+        ...
+
+    @overload
+    def addErrback(
+        self,
+        errback: Callable[Concatenate[Failure, _P], _NextDeferredResultT],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
+        ...
+
+    def addErrback(
+        self,
+        errback: Callable[..., Any],
+        *args: Any,
+        **kwargs: Any,
     ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
         """
         Convenience method for adding just an errback.
 
         See L{addCallbacks}.
         """
-        # type note: passthru constrains the type of errback in a way which mypy
-        #     can't propagate through to _NextDeferredResultT, so we have to
-        #     ignore a type error.
         return self.addCallbacks(
             passthru,
-            errback,  # type: ignore[arg-type]
+            errback,
             errbackArgs=args,
             errbackKeywords=kwargs,
         )
 
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[Concatenate[Union[_DeferredResultT, Failure], _P], Failure],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[
+            Concatenate[Union[_DeferredResultT, Failure], _P],
+            Deferred[_NextDeferredResultT],
+        ],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
     def addBoth(
         self,
         callback: Callable[
-            ...,
-            "Union[_NextDeferredResultT, Deferred[_NextDeferredResultT]]",
+            Concatenate[Union[_DeferredResultT, Failure], _P],
+            Union[_DeferredResultT, Failure],
         ],
-        *args: object,
-        **kwargs: object,
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
```

Comment:
I don't understand this annotation.  The combination of Union and `@overload` (which are both basically different ways to enlarge the type of the function) probably doesn't help.

This says "Give me a `Deferred[_DeferredResultT]` and a function that returns either `_DeferredResultT` or `Failure` and I will give you a `Deferred[_NextDeferredResultT]`".  `_NextDeferredResultT` isn't even bound in this statement so there's no way this could work.

I don't think this `@overload` is covering a meaningful case at all and should just be removed.

Reply:
OK, I think I have eliminated all the "callback callable returns `_DeferredResultT`" overloads. In order to make this easier to read, I've renamed the variables to `_SelfResultT` and `_NextResultT`, to have a bit less confusing repetition and hopefully make this type of mistake easier to spot.